TMS.Net07.Homework.Calculator/TMS.Net07.Homework.Calculator.Additional/Program.cs

```diff
@@ -0,0 +1,142 @@
+        {
+            Console.WriteLine(
+            $@"Availables algoritms:
+                1. {nameof(GetFactorialLoop)}
+                2. {nameof(GetFactorialRecursive)}
+                3. {nameof(GetFibonacciRecursive)}
+                4. {nameof(GetFibonacciRecursiveEffective)}");
+
+            while (true)
+            {
+                string[] inputValues = GetInputValues();
+                Console.WriteLine(GetResultMessage(inputValues));
+                Console.WriteLine("Try again? (y/n):");
+                if (!Console.ReadLine().Trim().ToLower().StartsWith('y'))
+                {
+                    break;
+                }
+            }
+        }
+
+        private static string[] GetInputValues()
+        {
+            string[] messages =
+            {
+            "Please, select algoritm (1-4):",
+            "Enter positive integer value:"
+            };
+            string[] inputValues = new string[messages.Length];
+            for (int i = 0; i < messages.Length; i++)
+            {
+                Console.WriteLine(messages[i]);
+                inputValues[i] = Console.ReadLine().Trim().ToLower();
+            }
+            return inputValues;
+        }
+
+        private static string GetResultMessage(string[] inputValues)
+        {
+            if (!int.TryParse(inputValues[0], out int idAlgoritm) ||
+                !int.TryParse(inputValues[1], out int number) || number < 0)
+            {
+                return "You entered is not valid integer value!";
+            }
+            decimal result;
+            try
+            {
+                switch (idAlgoritm)
+                {
+                    case 1:
+                        result = GetFactorialLoop(number);
+                        break;
+                    case 2:
+                        result = GetFactorialRecursive(number);
+                        break;
+                    case 3:
+                        result = GetFibonacciRecursive(number);
+                        break;
+                    case 4:
+                        result = GetFibonacciRecursiveEffective(number, new decimal[number + 1]);
+                        break;
+                    default:
+                        return "Algoritm is not available!";
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.Message);
+                result = -1;
+            }
+            return "Result: " + $"{(result < 0 ? "Error!" : result.ToString())}";
+        }
+
+        private static decimal GetFactorialLoop(int number)
+        {
+            if (number == 0) return 1;
+            decimal factorial = number;
+            try
+            {
+                for (int i = 1; i < number; i++)
+                {
+                    factorial = checked(factorial * i);
+                }
+            }
+            catch (Exception)
+            {
+                throw;
```

Comment:
It doesn't make sense to catch and then re-throw exception - the place, where it occured will be displayed in stack-trace.
However, there is one hidden stone:
if you would make:
```
            catch (Exception ex)
            {
                throw ex;
``` 
that would overwrite the initial stack trace and we would think, that exception were generated here.
That's why avoid re-throwing exception if we don't do anything with it in the  `catch` statement